docs/arch/adr-012-representing-change-history.md

```diff
@@ -0,0 +1,130 @@
+
+### Implementation plan
+
+1) Update any queries that join `change_notes` and `editions` to fully qualify `document_id` and `user_facing_version`
+   columns (i.e. use `"editions"."document_id"` not just `"document_id"`) to avoid ambiguous queries once we add columns
+2) Deploy `publishing-api` to all environments
+3) Add the `document_id` and `user_facing_version` columns to the `change_notes` table in a database migration
+4) Add an index to the `change_notes.document_id` column
+5) Update the code to ensure that whenever we save `ChangeNote` models we populate the `document_id` and
+   `user_facing_version` columns with the values from the referenced edition
+6) Populate the new columns in `change_notes` with a rake task (or some other kind of job)
+7) Update the `change_notes_for_edition` and `change_notes_for_linked_content_blocks` queries to avoid joining `editions`
+8) Hopefully, see a big performance improvement in change history queries!
+
+## Consequences
+
+### Queries to get change notes for a document will be much faster
+
+🚀
+
+Thanks to all the columns of interest being in the same table, we'll be able to do a much simpler index scan.
+
+### It will be easier for us to refactor the way we store superseded editions
+
+Publishing API's database is huge, largely because we store every historical version of every edition.
+
+Historical versions of editions are currently stored in the `editions` table, with the `content_store` column set to
+`NULL` and the `state` column set to `'superseded'`.
+
+One reason we need these superseded editions is that we need to look up change notes for documents, so we need the
+mapping from `document_id` to `edition_id`. Once we have the `document_id` column in `change_notes`, we don't need to
+query editions when getting change history for a document - we only need to look at the `change_notes` table.
+
+The fewer places where publishing-api relies on the presence of superseded editions in the editions table, the easier it
+will be to refactor the schema (e.g. by moving them into a separate `superseded_editions` table, or by partitioning the
+`editions` table on `content_store`).
+
+### We'll be storing the document to edition mapping in two places
+
+We're storing "which edition belongs to which document" in two places - the editions table, and the change notes table.
+
+This denormalization simplifies the joins required when querying change notes, but adds a bit of data redundancy.
+
+The main risk with data redundancy is update anomalies - for example, if we allowed editions to change which document
+they belonged to, it would be possible to update the document for an edition in the `editions` table but forget to
+update the other `document_id` in the `change_notes` table. In practice, editions never move from one document to
+another, and `user_facing_version` never changes once an edition is created. So I don't think there's a realistic data
+hazard in doing it this way.
```

Comment:
I agree with this for another reason. If we allowed editions to change which document they belonged to, the concept of having a sequence of change notes would be broken anyway. An individual change note relates to an edition, but a sequence of change notes relates to a document.